src/main/java/com/minecolonies/coremod/client/render/worldevent/ColonyBlueprintRenderer.java

```diff
@@ -159,96 +223,110 @@ static void renderBoxes(final WorldEventContext ctx)
+        final BlueprintPreviewData blueprintData = blueprintDataCache.getUnchecked(data.blueprint());
+        final Blueprint localBlueprint = blueprintData.getBlueprint();
+        if (localBlueprint == null)
+        {
+            return null;
+        }
 
-                    localBlueprint.setRenderSource(data.pos());
-                    if (data.boxOnly())
-                    {
-                        blueprintCache.put(data.pos(), new RenderData(null, box));
-                    }
-                    else
-                    {
-                        blueprintCache.put(data.pos(), new RenderData(blueprintPreviewData, box));
-                    }
+        if (blueprintData.isEmpty())
+        {
+            return new BoxRenderData(null, 0);
+        }
+        else
+        {
+            final BlockPos primaryOffset = localBlueprint.getPrimaryBlockOffset();
+            final BlockPos boxStartPos = data.pos().subtract(primaryOffset);
+            final BlockPos boxEndPos = boxStartPos.offset(localBlueprint.getSizeX() - 1, localBlueprint.getSizeY() - 1, localBlueprint.getSizeZ() - 1);
+            final BoxPreviewData box = new BoxPreviewData(boxStartPos, boxEndPos, data.hasAnchor() ? Optional.of(data.pos()) : Optional.empty());
 
-                    // only process one real blueprint per call
-                    break;
-                }
-            }
-            catch (final InterruptedException | ExecutionException ex)
+            return new BoxRenderData(box, data.builder());
+        }
+    }
+
+    private static void processPendingBlueprints()
+    {
+        final Iterator<Map.Entry<BlockPos, PendingRenderData>> iterator = pendingBoxes.entrySet().iterator();
+        while (iterator.hasNext())
+        {
+            final Map.Entry<BlockPos, PendingRenderData> entry = iterator.next();
+            final BoxRenderData box = tryLoadBox(entry.getValue());
+            if (box != null)
             {
-                //Noop
+                if (box.box() != null || box.builder() != 0)
+                {
+                    boxRenderCache.put(entry.getKey(), box);
+                }
+                iterator.remove();
             }
         }
     }
 
+    private static @NotNull BlueprintPreviewData makeBlueprintPreview(@NotNull final BlueprintCacheKey key)
+    {
+        final Future<Blueprint> blueprintFuture = StructurePacks.getBlueprintFuture(key.packName(), key.path());
+
+        final BlueprintPreviewData blueprintPreviewData = new BlueprintPreviewData();
+        blueprintPreviewData.setBlueprintFuture(blueprintFuture);
+        blueprintPreviewData.setPos(BlockPos.ZERO);
+        if (key.orientation().mirror() != Mirror.NONE)
+        {
+            blueprintPreviewData.mirror();
+        }
+        blueprintPreviewData.rotate(key.orientation().rotation());

```

Comment:
BTW @Nightenom, while you're here: this looks like a gap in the Structurize API, that I can't just set an existing `RotationMirror` value but have to break it into its components.

Also, it would be nice if there was a setter that does **not** trigger `syncChangesToServer()`, because that's completely pointless in this case.  Perhaps the caller should be doing that where needed instead of putting it in that class?

Reply:
gap is intentional, will break at major, will cause quite huge pr so I delayed it

for the second we can have boolean in ctor which may disable syncing permanently for that instance (I wouldnt do enable/disable methods bcs the decision needs to know context of origin)